File: backend/app/api/posts.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
from app.core.database import get_db
from app.models.models import Post, PostType, UserRole
from app.schemas.schemas import PostCreate, PostOut
from app.services.permission import get_current_user, check_admin

router = APIRouter(prefix="/posts", tags=["posts"])

@router.get("/", response_model=List[PostOut])
def get_posts(type: Optional[PostType] = None, db: Session = Depends(get_db)):
    query = db.query(Post)
    if type:
        query = query.filter(Post.type == type)
    return query.order_by(Post.is_pinned.desc(), Post.created_at.desc()).all()

@router.post("/", response_model=PostOut)
def create_post(post_in: PostCreate, current_user=Depends(get_current_user), db: Session = Depends(get_db)):
    # 权限检查：公告栏仅限管理员发帖
    if post_in.type == PostType.announcement and current_user.role != UserRole.admin:
        raise HTTPException(status_code=403, detail="Only administrators can post announcements")
    
    new_post = Post(
        content=post_in.content,
        type=post_in.type,
        is_anonymous=post_in.is_anonymous,
        user_id=current_user.id
    )
    db.add(new_post)
    db.commit()
    db.refresh(new_post)
    return new_post

@router.delete("/{post_id}")
def delete_post(post_id: int, current_user=Depends(get_current_user), db: Session = Depends(get_db)):
    post = db.query(Post).filter(Post.id == post_id).first()
    if not post:
        raise HTTPException(status_code=404, detail="Post not found")
    
    # 权限检查：管理员可删任何帖，用户仅可删自己的帖
    if current_user.role != UserRole.admin and post.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized to delete this post")
    
    db.delete(post)
    db.commit()
    return {"detail": "Post deleted successfully"}

File: backend-node/src/api/posts.js
const express = require('express');
const router = require('express').Router();
const pool = require('../core/database');
const { authenticate, authorize } = require('../middlewares/auth.middleware');

router.get('/', async (req, res) => {
  const type = req.query.type || 'announcement';
  try {
    const [posts] = await pool.query('SELECT * FROM posts WHERE type = ? ORDER BY is_pinned DESC, created_at DESC', [type]);
    res.json(posts);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.post('/', authenticate, async (req, res) => {
  const { content, type, is_anonymous } = req.body;
  const userId = req.user.id;
  try {
    const [result] = await pool.query(
      'INSERT INTO posts (user_id, content, type, is_anonymous) VALUES (?, ?, ?, ?)',
      [userId, content, type, is_anonymous || false]
    );
    res.status(201).json({ id: result.insertId, message: 'Post created' });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.delete('/:id', authenticate, async (req, res) => {
  const postId = req.params.id;
  const userId = req.user.id;
  const userRole = req.user.role;

  try {
    const [posts] = await pool.query('SELECT user_id FROM posts WHERE id = ?', [postId]);
    if (posts.length === 0) return res.status(404).json({ error: 'Post not found' });

    const post = posts[0];
    if (userRole !== 'admin' && post.user_id !== userId) {
      return res.status(403).json({ error: 'Forbidden: You can only delete your own posts' });
    }

    await pool.query('DELETE FROM posts WHERE id = ?', [postId]);
    res.json({ message: 'Post deleted successfully' });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;
